File: example/xx_three/loader_ply.cpp
module;
#include <infra/Cpp20.h>
#include <xx_three/ex.h>
module two.xxthree;

using namespace two;

#define CUSTOM_RENDER 0

EX(xx_loader_ply)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
#endif

	static ImporterPLY importer_ply = { app.m_gfx };

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 35.f; camera.m_near = 1.f; camera.m_far = 15.f;
		camera.m_eye = vec3(3.f, 0.15f, 3.f);
		camera.m_target = vec3(0.f, -0.1f, 0.f);

		viewer.m_viewport.m_to_gamma = true;

		//Colour colour = to_linear(rgb(0x72645b));
		Colour colour = to_colour(pow(to_vec3(rgb(0x72645b)), vec3(2.f)));

		viewer.m_viewport.m_clear_colour = colour;

		Zone& env = scene.m_env;
		env.m_background.m_colour = colour;

		env.m_fog.m_enabled = true;
		env.m_fog.m_colour = colour;
		env.m_fog.m_depth_begin = 2.f;
		env.m_fog.m_depth_end = 15.f;

		auto add_light = [&](vec3 d, Colour color, float intensity, bool shadows)
		{
			const quat r = look_dir(normalize(d));
			Node3& n = gfx::nodes(scene).add(Node3(vec3(0.f), r));
			Light& l = gfx::lights(scene).add(Light(n, LightType::Direct, shadows, color, intensity));
			//l.m_shadow_range = 4.f;
			//l.shadow.bias = -0.001;
		};

		Program& three = *app.m_gfx.programs().file("pbr/three");
		Program& phong = *app.m_gfx.programs().file("pbr/phong");

		// ground

		Material& groundmat = app.m_gfx.materials().create("plyground", [&](Material& m) {
			m.m_program = &phong;
			m.m_phong.m_diffuse = rgb(0x999999);
			m.m_phong.m_specular = rgb(0x101010);
		});

		Model& ground = app.m_gfx.shape(Rect(vec2(0.f), vec2(40.f)));
		Node3& n = gfx::nodes(scene).add(Node3(vec3(0.f, -0.5f, 0.f)));
		gfx::items(scene).add(Item(n, ground, 0, &groundmat));

		Material& material = app.m_gfx.materials().create("ply", [&](Material& m) {
			m.m_program = &three;
			m.m_base.m_flat_shaded = true;
			m.m_pbr.m_albedo = rgb(0x0055ff);
			// those are defaults in three.js standard material
			m.m_pbr.m_roughness = 0.5f;
			m.m_pbr.m_metallic = 0.5f;
		});

		// PLY models

		Model& dolphin = *app.m_gfx.models().file("dolphins"); // .ply
		Model& lucy = *app.m_gfx.models().file("Lucy100k"); // .ply
			
		Node3& n0 = gfx::nodes(scene).add(Node3(vec3(-0.2f, 0.f, 0.3f), quat(vec3(-c_pi2, 0.f, 0.f)), vec3(0.001f)));
		gfx::items(scene).add(Item(n0, dolphin, 0, &material));

		Node3& n1 = gfx::nodes(scene).add(Node3(vec3(-0.2f, -0.02f, -0.2f), quat(vec3(0.f)), vec3(0.0006f)));
		gfx::items(scene).add(Item(n1, lucy, 0, &material));
	
		// lights

		scene.m_env.m_skylight = { true, 1.f, -y3, vec3(0.f), rgb(0x443333), rgb(0x111122) };

		add_light(normalize(vec3(-1.f, -1.f, -1.f)), rgb(0xffffff), 1.35f, true);
		add_light(normalize(vec3(-0.5f, -1.f, 1.f)), rgb(0xffaa00), 1.f, true);
	}

	const float timer = app.m_gfx.m_time * 0.5f;

	Camera& camera = viewer.m_camera;
	camera.m_eye.x = sin(timer) * 2.5f;
	camera.m_eye.z = cos(timer) * 2.5f;
}

File: example/xx_three/two/loader_ply.js
// loader_ply.js

var viewer = two.ui.scene_viewer(panel);
//two.ui.orbit_controls(viewer);

var scene = viewer.scene;

if(init) {
    this.importerPLY = new two.ImporterPLY(app.gfx);
    
    var camera = viewer.camera;
    camera.fov = 35.0; camera.near = 1.0; camera.far = 15.0;
    camera.eye = new two.vec3(3.0, 0.15, 3.0);
    camera.target = new two.vec3(0.0, -0.1, 0.0);

    var bg = two.to_linear(two.rgb(0x72645b));
    bg = new two.Colour(bg.r, bg.g, bg.b);
    
    viewer.viewport.to_gamma = true;
    viewer.viewport.clear_colour = bg;
    
    var env = scene.env;
    env.background.colour = bg;
    
    env.fog.enabled = true;
    env.fog.colour = bg;
    env.fog.depth_begin = 2.0;
    env.fog.depth_end = 15.0;
    
    var zeroq = new two.quat(new two.vec3(0.0));

    function add_light(scene, r, color, intensity, shadows) {

        var n = scene.nodes().add(new two.Node3(new two.vec3(0.0), r));
        var l = scene.lights().add(new two.Light(n, two.LightType.Direct, shadows, color, intensity));
        //l.shadow_range = 4.0;
        //l.shadow.bias = -0.001;
    }

    var three = app.gfx.programs.fetch('pbr/three');
    var phong = app.gfx.programs.fetch('pbr/phong');

    // Ground

    var groundmat = app.gfx.materials.create('ground'); var m = groundmat;
        m.program = phong;
        m.phong.diffuse.value = two.rgb(0x999999);
        m.phong.specular.value = two.rgb(0x101010);
        // Phong

    var ground = app.gfx.shape(new two.Rect(new two.vec2(0.0), new two.vec2(40.0)));
    var n = scene.nodes().add(new two.Node3(new two.vec3(0.0, -0.5, 0.0), zeroq));
    scene.items().add(new two.Item(n, ground, 0, groundmat));

    var material = app.gfx.materials.create('ply'); m = material;
        m.program = three;
        m.base.flat_shaded = true;
        m.pbr.albedo.value = two.rgb(0x0055ff);
        m.pbr.roughness.value = 0.5;
        m.pbr.metallic.value = 0.5;

    // PLY file

    var dolphin = app.gfx.models.file('dolphins'); // .ply
    
    var ndolphin = scene.nodes().add(new two.Node3(new two.vec3(-0.2, 0.0, 0.3), new two.quat(new two.vec3(-Math.PI / 2, 0.0, 0.0)), new two.vec3(0.001)));
    scene.items().add(new two.Item(ndolphin, dolphin, 0, material));

    var lucy = app.gfx.models.file('Lucy100k'); // .ply

    var nlucy = scene.nodes().add(new two.Node3(new two.vec3(-0.2, -0.02, -0.2), zeroq, new two.vec3(0.0006)));
    scene.items().add(new two.Item(nlucy, lucy, 0, material));

    // Lights

    var skylight = scene.env.skylight;
    skylight.enabled = true;
    skylight.intensity = 1.0;
    skylight.position = new two.vec3(0.0, 1.0, 0.0);
    skylight.color = two.rgb(0x443333);
    skylight.ground = two.rgb(0x111122);

    add_light(scene, new two.quat(-0.325, 0.325, 0.0, 0.888), two.rgb(0xffffff), 1.35, true);
    add_light(scene, new two.quat(-0.816, 0.408, 0.0, 0.408), two.rgb(0xffaa00), 1.0, true);
    
    //add_light(scene, new two.vec3(-1.0, -1.0, -1.0), two.rgb(0xffffff), 1.35, true);
    //add_light(scene, new two.vec3(-0.5, -1.0, 1.0), two.rgb(0xffaa00), 1.0, true);
}

var timer = app.gfx.time * 0.5;

var camera = viewer.camera;
camera.eye.x = Math.sin(timer) * 2.5;
camera.eye.z = Math.cos(timer) * 2.5;
